Code contracts: do not interleave checking and instrumenting

tautschnig opened the pull request:
Loop normalisation must first check that all loops described via `natural_loopst` are of the expected shape and then start inserting or transforming instructions. Else, and depending on loop shapes and loop nesting, the instruction iterators held in the `natural_loopst` may no longer be adequate descriptions of the loops. (The iterators weren't invalidated, but would no longer point to heads or loop ends.)

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Code contracts: do not interleave checking and instrumenting

Loop normalisation must first check that all loops described via
`natural_loopst` are of the expected shape and then start inserting or
transforming instructions. Else, and depending on loop shapes and loop
nesting, the instruction iterators held in the `natural_loopst` may no
longer be adequate descriptions of the loops. (The iterators weren't
invalidated, but would no longer point to heads or loop ends.)

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -914,6 +920,42 @@ void code_contractst::apply_loop_contract(
+    // Check 1. (i \in loop_content) ==> loop_head <= i <= loop_end
+    for(const auto &i : loop_content)
+    {
+      if(
+        loop_head->location_number > i->location_number ||
+        i->location_number > loop_end->location_number)
```
Could we add a regression test to cover this case somehow?

tautschnig replied:
I'll try to produce a test, discussion with @remi-delmas-3000 suggests that nested do-while loops are most likely to demonstrate the failing behaviour.

tautschnig replied:
I have added a test with nested loops, which demonstrates yet another type of problem (see comment in nested.desc), but I'll leave that for another PR.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -914,6 +920,42 @@ void code_contractst::apply_loop_contract(
+    // Since we perform assigns clause instrumentation by sequentially
+    // traversing instructions from `loop_head` to `loop_end`,
+    // here we ensure that all instructions in `loop_content` belong within
+    // the [loop_head, loop_end] target range
```
```suggestion
    // the [loop_head, loop_end] target range.
```

feliperodri reviewed (approved):
LGTM (apart from the CI failures).

peterschrammel commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -914,6 +920,42 @@ void code_contractst::apply_loop_contract(
       throw 0;
     }
 
+    // By definition the `loop_content` is a set of instructions computed
```
Isn't that usually called "loop body"?

tautschnig committed:
Restrict function pointers: update goto program

Maintain goto-program invariants when modifying a program via
function-pointer restrictions.

tautschnig committed:
Rename loop_content to loop_body

This is to use standard terminology.
